presto-main/src/main/java/com/facebook/presto/ExceededMemoryLimitException.java

```diff
@@ -47,19 +55,36 @@ public static ExceededMemoryLimitException exceededLocalBroadcastMemoryLimit(Dat
                 format("Query exceeded per-node broadcast memory limit of %s [%s]", maxMemory, additionalFailureInfo));
     }
 
-    public static ExceededMemoryLimitException exceededLocalTotalMemoryLimit(DataSize maxMemory, String additionalFailureInfo)
+    public static ExceededMemoryLimitException exceededLocalTotalMemoryLimit(
+            DataSize maxMemory,
+            String additionalFailureInfo,
+            boolean heapDumpOnExceededMemoryLimitEnabled,
+            Optional<String> heapDumpFilePath)
     {
+        performHeapDumpIfEnabled(heapDumpOnExceededMemoryLimitEnabled, heapDumpFilePath);
         return new ExceededMemoryLimitException(EXCEEDED_LOCAL_MEMORY_LIMIT,
                 format("Query exceeded per-node total memory limit of %s [%s]", maxMemory, additionalFailureInfo));
     }
 
-    public static ExceededMemoryLimitException exceededLocalRevocableMemoryLimit(DataSize maxMemory, String additionalFailureInfo)
+    public static ExceededMemoryLimitException exceededLocalRevocableMemoryLimit(
+            DataSize maxMemory,
+            String additionalFailureInfo,
+            boolean heapDumpOnExceededMemoryLimitEnabled,
+            Optional<String> heapDumpFilePath)
     {
+        performHeapDumpIfEnabled(heapDumpOnExceededMemoryLimitEnabled, heapDumpFilePath);
         return new ExceededMemoryLimitException(
                 EXCEEDED_REVOCABLE_MEMORY_LIMIT,
                 format("Query exceeded per-node revocable memory limit of %s [%s]", maxMemory, additionalFailureInfo));
     }
 
+    private static void performHeapDumpIfEnabled(boolean heapDumpOnExceededMemoryLimitEnabled, Optional<String> heapDumpFilePath)
+    {
+        if (heapDumpOnExceededMemoryLimitEnabled && heapDumpFilePath.isPresent()) {
```

Comment:
Does it make sense to check whether the heapDumpFilePath is a valid path?

Reply:
This is a session property per query and not a configuration at startup. I don't think it is helpful to check here since it will fail if we are wrong, but it might make sense to do something like https://github.com/prestodb/presto/blob/master/presto-main/src/main/java/com/facebook/presto/spiller/FileSingleStreamSpillerFactory.java#L101 ?

I am torn on whether the query should fail fast at startup if the file can't be created vs just letting it fail to create the heap dump. It's a bit finicky because we would need to check for each query whether the file can be created.

This is for debug so the temptation is to just do what is simple.

Reply:
I did thought of adding some kind of valid path check but did not pursued on it since this is more of a debugging feature. If heapDump is enabled and heapDumpPath is not specified, a tmp path is automatically constructed for heapDump.  Users have the ability to override the heapDumpPath, if needed and they are expected to provide a valid path. 

Also, the heapDump is a best effort feature and will not fail the query on exceptions. If any exception is thrown, it will be swallowed and query execution will fail with proper EXCEEDED_MEMORY_LIMIT exception.